PEP 661: Proposed typing specification changes (#4065)
python/peps: peps/pep-0661.rst

Ordering comparisons are undefined for sentinel objects.

Typing
------

> To make usage of sentinels convenient in typed Python code, we propose to

Reviewer: In my view this is about much more than convenience: Not having clear type annotations is one of the 3 main deficiencies of existing implementations of sentinels, as outlined in the "Motivation" sections.

The authors of this PR have much more experience than I regarding typing in Python, but I think at this point we can agree that typing considerations are more than an "afterthought" for convenience.

As a simple change, perhaps we could use "clear and simple" rather than "convenient" here?

Author: ```suggestion
To make usage of sentinels clear and simple in typed Python code, we propose to
```
Sure, that works too.

> To make usage of sentinels convenient in typed Python code, we propose to add

Reviewer 2: Typo: remove "add"

amend the type system with a special case for sentinel objects.

> Sentinel objects may be used in :term:`type expressions <typing:type expression>`,
> representing themselves.
> This is similar to how ``None`` is handled in the existing type system.
> For example::

Reviewer: Formatting nit:
```suggestion
Sentinel objects may be used in
:term:`type expressions <typing:type expression>`, representing themselves.
This is similar to how ``None`` is handled in the existing type system. For
example::
```

    from sentinels import Sentinel

    MISSING = Sentinel('MISSING')

    def foo(value: int | MISSING = MISSING) -> int:
        ...

More formally, type checkers should recognize sentinel creations of the form

> ``NAME = Sentinel('NAME')`` as creating a new sentinel objects. If the name

Reviewer 2: Typo: "a new sentinel objects" should be "a new sentinel object" (singular).

Author: Thanks, fixed both

passed to the ``Sentinel`` constructor does not match the name the object is
assigned to, type checkers should emit an error.

Sentinels defined using this syntax may be used in
:term:`type expressions <typing:type expression>`. They

> represent a :term:`fully static type <typing:fully static type>` that has a single member, the sentinel object itself.

Reviewer: ```suggestion
represent a :term:`fully static type <typing:fully static type>` that has a
single member, the sentinel object itself.
```
